Mage.Sets/src/mage/cards/e/EcologicalAppreciation.java

```diff
@@ -0,0 +1,150 @@
+package mage.cards.e;
+
+import mage.MageObject;
+import mage.abilities.Ability;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.ExileSpellEffect;
+import mage.cards.*;
+import mage.constants.CardType;
+import mage.constants.Outcome;
+import mage.constants.Zone;
+import mage.filter.FilterCard;
+import mage.filter.StaticFilters;
+import mage.filter.common.FilterCreatureCard;
+import mage.game.Game;
+import mage.players.Player;
+import mage.target.TargetCard;
+import mage.target.common.TargetOpponent;
+
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ *
+ * @author htrajan
+ */
+public final class EcologicalAppreciation extends CardImpl {
+
+    public EcologicalAppreciation(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.SORCERY}, "{X}{2}{G}");
+
+        // Search your library and graveyard for up to four creature cards with different names that each have mana value X or less and reveal them. An opponent chooses two of those cards. Shuffle the chosen cards into your library and put the rest onto the battlefield. Exile Ecological Appreciation.
+        this.getSpellAbility().addEffect(new EcologicalAppreciationEffect());
+        this.getSpellAbility().addEffect(ExileSpellEffect.getInstance());
+    }
+
+    private EcologicalAppreciation(final EcologicalAppreciation card) {
+        super(card);
+    }
+
+    @Override
+    public EcologicalAppreciation copy() {
+        return new EcologicalAppreciation(this);
+    }
+}
+
+class EcologicalAppreciationEffect extends OneShotEffect {
+
+    EcologicalAppreciationEffect() {
+        super(Outcome.Benefit);
+        staticText = "search your library and graveyard for up to four creature cards with different names that each have mana value X or less and reveal them. An opponent chooses two of those cards. Shuffle the chosen cards into your library and put the rest onto the battlefield";
+    }
+
+    private EcologicalAppreciationEffect(final EcologicalAppreciationEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public EcologicalAppreciationEffect copy() {
+        return new EcologicalAppreciationEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        if (player == null) {
+            return false;
+        }
+        int xValue = source.getManaCostsToPay().getX();
+        TargetCard targetCardsInLibrary = new EcologicalAppreciationTarget(Zone.LIBRARY, 4, xValue);
+        targetCardsInLibrary.setNotTarget(true);
+        boolean searched = player.choose(Outcome.PutCreatureInPlay, new CardsImpl(player.getLibrary().getCards(game)), targetCardsInLibrary, game);
+        Cards cards = new CardsImpl(targetCardsInLibrary.getTargets());
+
+        if (cards.isEmpty()) {
+            if (searched) {
```

Comment:
It's a smell code to call shuffle three times. You can store it the status var and use it at the end. Also don't use `searched` at all -- always shuffle it. Cause cheating possible: user can open choose dialog and disconnect (see #4263) -- so you will get a false here, but he already see the cards order.

Reply:
fixed